File: src/typescript/src/xiaoyi/domain/token/primitive/string.ts
/**
 * # String Primitive
 *
 * `string` provides UTF-8 string type with encoding validation.
 *
 * Path: `xiaoyi.domain.token.primitive.string`
 *
 * - Layer 0: `domain`
 * - Layer 1: `token`
 * - Layer 2: `primitive`
 * - Layer 3: `string`
 *
 * @module domain.token.primitive.string
 * @brief UTF-8 string primitive
 * @group Domain
 * @since 0.1.0
 * @author Miruamel
 * @see domain.token.primitive
 * @see domain.token.primitive.int
 */

/** String type alias (owned UTF-8). */
export type String = string;

/** String slice type alias (borrowed UTF-8). */
export type Str = string;

/**
 * Create new empty string.
 *
 * @returns Empty String
 * @since 0.1.0
 */
export function newString(): String {
  return "";
}

/**
 * Create string from string slice.
 *
 * @param s - String slice
 * @returns Owned String
 * @since 0.1.0
 */
export function fromString(s: Str): String {
  return s;
}

/**
 * Check if string is valid UTF-8.
 *
 * @param bytes - Byte array
 * @returns true if valid UTF-8
 * @since 0.1.0
 */
export function isValidUtf8(bytes: Uint8Array): boolean {
  try {
    new TextDecoder("utf-8", { fatal: true }).decode(bytes);
    return true;
  } catch {
    return false;
  }
}

/**
 * Get string length in characters (code points).
 *
 * @param s - String slice
 * @returns Character count
 * @since 0.1.0
 */
export function charLen(s: Str): number {
  return [...s].length;
}

File: src/python/xiaoyi/domain/token/primitive/string.py
"""
# String Primitive

`string` provides UTF-8 string type with encoding validation.

Path: `xiaoyi.domain.token.primitive.string`

- Layer 0: `domain`
- Layer 1: `token`
- Layer 2: `primitive`
- Layer 3: `string`

@module xiaoyi.domain.token.primitive.string
@brief UTF-8 string primitive
@group Domain
@since 0.1.0
@author Miruamel
@see xiaoyi.domain.token.primitive
@see xiaoyi.domain.token.primitive.int
"""

from typing import Any


#: String type alias (owned UTF-8).
String = str

#: String slice type alias (borrowed UTF-8).
Str = str


def new_string() -> String:
    """
    Create new empty string.

    @return Empty String
    @since 0.1.0
    """
    return ""


def from_string(s: Str) -> String:
    """
    Create string from string slice.

    @param s String slice
    @return Owned String
    @since 0.1.0
    """
    return s


def is_valid_utf8(data: bytes) -> bool:
    """
    Check if bytes is valid UTF-8.

    @param data Byte array
    @return True if valid UTF-8
    @since 0.1.0
    """
    try:
        data.decode("utf-8")
        return True
    except UnicodeDecodeError:
        return False


def char_len(s: Str) -> int:
    """
    Get string length in characters (code points).

    @param s String slice
    @return Character count
    @since 0.1.0
    """
    return len(s)


__all__ = [
    "String",
    "Str",
    "new_string",
    "from_string",
    "is_valid_utf8",
    "char_len",
]